climada/util/calibrate/impact_func.py

```diff
@@ -0,0 +1,225 @@
+"""Module for calibrating impact functions"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field, InitVar
+from typing import Callable, Mapping, Optional, Tuple, Union, Any, Dict, List
+from numbers import Number
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import (
+    Bounds,
+    LinearConstraint,
+    NonlinearConstraint,
+    minimize,
+)
+from bayes_opt import BayesianOptimization
+
+from climada.hazard import Hazard
+from climada.entity import Exposures, ImpactFunc, ImpactFuncSet
+from climada.engine import Impact, ImpactCalc
+
+
+def cost_func_rmse(
+    impact: Impact,
+    data: pd.DataFrame,
+    impact_proc: Callable[[Impact], pd.DataFrame] = lambda x: x.impact_at_reg(),
+) -> Number:
+    return np.sqrt(np.mean(((impact_proc(impact) - data) ** 2).to_numpy()))
+
+
+def impf_step_generator(threshold: Number, paa: Number) -> ImpactFuncSet:
+    return ImpactFuncSet(
+        [
+            ImpactFunc.from_step_impf(
+                haz_type="RF", intensity=(0, threshold, 100), paa=(0, paa)
+            )
+        ]
+    )
+
+
+ConstraintType = Union[LinearConstraint, NonlinearConstraint, Mapping]
+
+
+@dataclass
+class Input:
+    """Define the static input for a calibration task
+
+    Parameters
+    ----------
+    hazard : climada.Hazard
+        Hazard object to compute impacts from
+    exposure : climada.Exposures
+        Exposures object to compute impacts from
+    data : pandas.Dataframe
+        The data to compare computed impacts to. Index: Event IDs matching the IDs of
+        ``hazard``. Columns: Arbitrary columns.
```

Comment:
Is this not too restrictive to have it event-matching? It might be sensible to optimize for the total over many events for instance. So, if this is not adding too much complexity, I would allow for it (since the only important factor, at first sight, is the cost function matching the data).